### ISSUE TITLE
Option for ignoring original formatting 100%

### ISSUE THREAD

**Issue by ccorcos** [prettier/prettier#2068]
Edit by @lydell: Prettier currently keeps multiline object literals multiline even if they fit in a single line, and does special things to blank lines: removes some, collapses consecutive blank lines into one and never adds blank lines. This is because the majority of people seem to want this level of control since we haven’t found a super good heuristic for printing those things. Some people prefer a 100% uniform printing, though, that _completely_ ignores original formatting, even if it sometimes results in “uglier” code. We should discuss this.

Original post:

Is it possible to have a strict option that disambiguates whitespace and always prints out the same so long as the AST is the same? I'm running into some weird issues where lines wrap and then don't unwrap when they are under the line-length limit. This would greatly help with some of my merge conflicts right now :)

P.S. I'm using typescript.

Thanks

**Comment by vjeux** [comment]
You are probably thinking about objects that keep being expanded when there's a `\n` inside, even if they don't fit.

There are unfortunate cases like this where we can't just blindly print the ast. We try to keep them to a minimum though.

But, I think that the root of your issue is that the workflow for solving merge conflicts is not well oiled out. We should focus on making this better. If you could explain what you are currently doing and what are the merge conflicts and how you try to solve them, that would be super helpful.

**Comment by ccorcos** [comment]
Exactly. So here's a good example of something that is giving us trouble.

This was the code in branch1.

```ts
export const trackSelectTemplate = function(args: { name: string }) {
	analyticsHelpers.track("select_template", args)
}
```

And this is the same code on master.

```ts
export const trackSelectTemplate = function(
	args: {
		name: string
	}
) {
	analyticsHelpers.track("select_template", args)
}
```

Both have been run through prettier and they're stable. I honestly have no idea how we got into this situation though... It seems like a solid solution is just to unwrap lines that can be unwrapped. What's wrong with that solution?

**Comment by lydell** [comment]
@ccorcos See https://github.com/prettier/prettier/blob/master/Rationale.md#multi-line-objects and the issues it links to. I'm sorry but this has already been discussed a lot and the best trade-off we've come up with is to keep objects multliline if they have a newline inside.

**Comment by ccorcos** [comment]
Interesting. Would it be possible to have an option at least? It appears that the discussion has focused mostly on aesthetics but the reason I bring it up is to have an entirely unambiguous formatter. I don't actually care that much about how it looks. I just want one way of doing things.

**Comment by lydell** [comment]
@ccorcos How would you handle empty lines?

**Comment by ccorcos** [comment]
I think we could easily come up with some conventions, don't you?

- empty line after imports

```
import x from "x"

function blah() {}
```

- empty line around function definitions

```
const w = 10

function x() {}

function y() {}
```

- empty line before comments

```
const x = 10

// this does something
const y = 20
const z = 50
```

- empty line around class methods

Can you foresee any tricky scenarios?

**Comment by lydell** [comment]
I think this would be an interesting option so I’m opening up this for discussion for a while.

**Comment by SimenB** [comment]
I'd also love to see an option to ignore the style of the input. I'd think I'd not use it all the time, but would love to have the option. In the rare case where I find the result less readable, I imagine just reverting it and prettier won't touch it the next time.

**Comment by ccorcos** [comment]
Exactly. And I think when coding in a large organization, the less ambiguity the better :)

**Comment by lydell** [comment]
ESLint rule to take inspiration from when it comes to blank lines: http://eslint.org/docs/rules/padding-line-between-statements

**Comment by josephfrazier** [comment]
This is tangentially related, but I noticed that adding `--debug-check` to our test suites causes 11 of them to fail the `prettier(x) == prettier(prettier(x))` condition, and I'm guessing that it's because input formatting is taken into account. See here for details: https://github.com/prettier/prettier/issues/3512#issuecomment-352622275

**Comment by lydell** [comment]
`prettier(x) == prettier(prettier(x))` should _always_ be true, regardless if we look at the original input or not. Everything else is a bug.

On a related note, I often think about this issue, but have never had the time and energy to work on it yet.

**Comment by rattrayalex** [comment]
From @jlongster in https://github.com/prettier/prettier/issues/3524#issuecomment-353173008, referring to this issue:

> @rattrayalex I actually would use the hell out it, but I'm not sure about adding it as an option. Personally I'd like to keep this as zero config as we can get and minimize the amount of differences. While it's not as big of a change as the this issue, it's still a lot bigger change than just single/double quotes, bracket spacing, etc. I don't feel as strongly about it though and if other maintainers decide it should be an option, you can do it. I'd vote against but wouldn't block that decision.

I think that if we're going to collapse objects, we'll need really good heuristics about when we do. Identifying when something is a "config object" and best split out sounds both **tough** and **important**. Same with "what's a chuck of code", as @ccorcos mentioned. (hi Chet!)

Perhaps we could have a "temporary, experimental" option in prettier 1.x while we iterate on these heuristics, so we can gather feedback from the community? The goal would be to launch prettier 2.0 with "full input-text ignorance".

**Comment by rattrayalex** [comment]
Looking through the linked issues a bit (I see https://github.com/prettier/prettier/issues/74, https://github.com/prettier/prettier/issues/88, 

@vjeux had this idea in https://github.com/prettier/prettier/issues/88#issuecomment-275584603: 

> Another idea: automatically expand all the objects at the top level. This may be weird and feel inconsistent but would solve the problem of RN StyleSheet.create and the brunch config.

In other words, "if any node of a nested object must expand, all nodes must expand", which I think should extend to include arrays as well (eg; the top example in #74). There are likely to be subtleties of course 😄 

Thoughts

Also, @ccorcos are you interested in putting together a PR along the lines of what you described in 
https://github.com/prettier/prettier/issues/2068#issuecomment-307995925 ? (Assuming there's interest in proceeding along this path, which I think there is)

**Comment by lydell** [comment]
What about "if three or more items in the object, expand it"?

**Comment by rattrayalex** [comment]
@lydell what if the three items are all shorthand notation and have short names? eg; `runProgram({data, config, onChange})` should probably not expand. I think that might be a useful heuristic in other scenarios though.

(I'll be afk for the next 10 days; otherwise would give it a go).

**Comment by azz** [comment]
"if three or more **and** there's at least one non-shorthand property"?

```js
runProgram({data, config, onChange});
runProgram({
  data: null,
  config,
  onChange
});
runProgram({data: null, config: null});
```

(Probably not a good idea but I typed it up so I may as well post it)

**Comment by a7madgamal** [comment]
I couldn't agree more. Adding to the discussion, I don't have a specific issue to request this.
It's simply taking what prettier started and believed in to the next level. I had a lot of discussions in previous and current companies about silly single lines. Add one here, remove this one... 
Regardless of how legitimate or sane these requests. I would rather avoid them and focus on what matters. 

However, I also admit it built an appreciation for better code readability. Strict prettier mode or even better adobting this into the next prettier version (IMO) would save the millions of developers using prettier even more time and effort. 
Thanks!

**Comment by tannerlinsley** [comment]
I'm still interested in this. Any movement here?

**Comment by lydell** [comment]
No movement. I miss this every day.

**Comment by loynoir** [comment]
Can any one explain to me, why this marked as `difficulty: hard`, like I'm 5 years old?  🤔
Seem like kind of like
```
featPrettier = x => currentPrettier(minifier(x, every_no_aggressive_options))

// for every x, always
featPrettier(x) === featPrettier(featPrettier(x))
```

**Comment by alexander-akait** [comment]
@loynoir PR welcome if you think it is easy

**Comment by loynoir** [comment]
Current my acceptable workaround

```js
// Gulpfile.ts
import gulp from 'gulp';
import terser from 'gulp-terser';
import prettier from 'gulp-prettier';
// import { argv } from 'yargs'

if (process.env.prefix) {
  process.chdir(process.env.prefix)
}

function terserPrettier() {
  return gulp.src('dist/**/*js')
    .pipe(terser({
      compress: false,
      mangle: false,
      keep_classnames: true,
      keep_fnames: true,
      // // @ts-expect-error
      // keep_number: true,
      output: {
        keep_quoted_props: true,
      }
    }))
    .pipe(prettier())
    .pipe(gulp.dest('dist.min'));
}

terserPrettier.flags = {};

gulp.task('default', terserPrettier)
```

```sh
# diffing AST changes
$ meld a/dist b/dist
```

**Comment by loynoir** [comment]
@alexander-akait 
Sorry. Apologize if I offense. 🙇
I just curious what is the bottleneck here...

**Comment by uriva** [comment]
Hi I'm interested in the following subset of this problem:

prettier allows:

```
const newState = { ...state };
```

but also
```
const newState = { 
  ...state,
};
```

I think just choosing between these too would go a long way for me. was this discussed? if so what is the status?

thanks

**Comment by nullromo** [comment]
@uriva Unfortunately the maintainers of Prettier have rejected this idea many times over. They are concerned with [having as few options as possible](https://prettier.io/docs/en/option-philosophy.html) which means they won't add a flag for it, and also with [satisfying people who care about all their objects looking the same for some reason or another](https://github.com/prettier/prettier/issues/74#issue-199965534), which means they won't change this behavior. So while it would be nice if prettier were "deterministic" or "reversible," it's not going to happen, no matter [ [how](https://github.com/prettier/prettier/issues/14367) ]  [ [many](https://github.com/prettier/prettier/issues/12541) ]  [ [times](https://github.com/prettier/prettier/issues/11856) ] it's brought up. I agree that it's a shame that it's missing this feature, since I think Prettier is such a great and well-made tool, apart from it's lack of determinism.

**Comment by uriva** [comment]
I argue it should be the default.
Let's fork..

**Comment by pauldraper** [comment]
> I'm sorry but this has already been discussed a lot and the best trade-off we've come up with is to keep objects multliline if they have a newline inside.

No, it's more obscure than that.

The object is kept multiline *if there is a new line between the opening brace and first property*

The following are formatted as multi-line:

```js
const obj = {
  name1: "value1",
  name2: "value2",
};
```

```js
const obj = {
  name1: "value1", name2: "value2",
};
```

```js
const obj = {
  name1: "value1",
  name2: "value2", };
```

The following are formatted as single-line:

```js
const obj = { name1: "value1", name2: "value2", };
```

```js
const obj = { name1: "value1",
  name2: "value2",
};
```

The current Prettier documentation explicitly says

# THIS "IS NOT A FEATURE"

> The semi-manual formatting for object literals is in fact a workaround, not a feature.

**Comment by pauldraper** [comment]
You can use https://github.com/resolritter/prettier-plugin-compactify though that project isn't active. (And you can only have one of these "preprocess" plugins per language....I already use prettier-plugin-organize-imports.)

**Comment by roman-petrov** [comment]
I tried the plugin `prettier-plugin-compactify` and it working almost great, but a bit outdated: it does not supports some latest TypeScript features like import assertions.  Just hoping that such kind of preprocessor will be part of prettier core as a configuration options some days...

**Comment by nullromo** [comment]
@pauldraper Looks like a very well-reasoned PR. I wish you the best of luck!

**Comment by nullromo** [comment]
[Prettier 3.5.0](https://prettier.io/blog/2025/02/09/3.5.0) which released yesterday now includes @pauldraper's `objectWrap` option, which most likely closes this issue.

**Comment by uriva** [comment]
@nickserv function parameters whitespace (maybe with adding a comma) is also a redundant ambiguity imo

**Comment by fisker** [comment]
Close for now.
